### ISSUE TITLE
Django-nonrel should support cascading deletes

### ISSUE THREAD

**Issue by Wilfred** [django-nonrel/django-nonrel#1]
Following discussion on [Django #17337](https://code.djangoproject.com/ticket/17338) ([relevant mailing list thread](http://groups.google.com/group/django-developers/browse_thread/thread/a3d3c5d3bace09a6#msg_b2c0189ec4f7c588)) it looks like we should be able to fix this. Django does not actually depend on the DB for this.

The related object deletion seems to happen in `django.db.models.deletion.Collector.collect` but it needs more digging to see what we're missing.

**Comment by Wilfred** [comment]
In fact, I think simply removing [the patch](https://code.djangoproject.com/attachment/ticket/17338/supports-deleting-related.patch) would fix this.

**Comment by jonashaag** [comment]
@wkornewald maybe you could give some advice

**Comment by wkornewald** [comment]
Cascading deletes are a problem on App Engine because they make deletes in transactions impossible (non-pk queries aren't allowed within transactions). Also, they can be very inefficient at least on App Engine because they can cause a lot of additional queries. That's why we turned them off on NoSQL DBs, for now.

Our plan was to allow backends to at least move the cascading delete into a background task (i.e. stay compatible, but be more efficient on certain DBs). Although Django now allows to override the delete behavior I don't think that's the real solution because it makes App Engine code (and similarly affected DBs) less reusable with both SQL and other NoSQL DBs. It would be better if you could use exactly the same code on all DBs.

**Comment by Wilfred** [comment]
You're right, it doesn't work inside a transaction. However, this is making the common case of deleting outside a transaction much harder. I keep seeing non-rel projects where people have simply reimplemented Collector in a more buggy fashion.

I propose that `Model.delete` and `QuerySet.delete` should have a keyword argument of `delete_related`. This way you can still manually delete things if you're in an App Engine transaction, but otherwise you can use Django's normal helpfulness.

Correct me if I'm wrong, but I don't think Django on App Engine would do any more queries than vanilla Django does in order to delete related models. Again, if performance becomes an issue then allowing the user to request related objects to be left well alone would give them flexibility.

**Comment by wkornewald** [comment]
The biggest concern for me is that the delete_related parameter doesn't solve the problem of platform independence. (a) it wouldn't work with SQL due to consistency errors and (b) most of the code out there would be written for SQL and non-App Engine NoSQL DBs, so the delete_related parameter would be almost exclusively used by App Engine developers because almost nobody else needs it. That makes code reuse difficult for GAE people.

Additionally, Django would be less efficient at deleting related objects because it doesn't run queries in parallel and it can't do batch deletes across multiple tables.

The ideal solution would be to allow the backend to move the deletion into a background task because that gives us (1) portable code that (2) runs with maximum efficiency and (3) a responsive user experience (the user doesn't have to wait for all the queries to be executed). Well, maybe I'm just over-exaggerating and the delete_related parameter is sufficient because you rarely have a problem with the cascade of extra queries. If you decide to go with the delete_related parameter I'd suggest that you at least test with Django-nonrel how it works out for you and other users in practice before you submit the patch to the Django team. If nobody complains it's probably fine. ;)

**Comment by Wilfred** [comment]
Patch provided works in most situations, but there are certain circumstances where it produces a JOIN query. So Waldemar is right as usual :). To fix the JOIN will require us to iterate over stuff and do more queries than vanilla Django.

I've not got to the bottom of it yet. I can only reproduce it on a site using nonrel_search so more digging is required.

**Comment by wrwrwr** [comment]
I've been playing with the cascade deletes reenabled today (just the first part of your changes) and also hit the "database does not support joins" exception -- with on_delete=models.SET_NULL. Nevertheless other than that all seems to work fine (neither transactions nor efficiency should be an issue in my small case and getting rid of all delete patching was good).

**Comment by jonashaag** [comment]
@wrwrwr can you drop me an email or send a message on GitHub? (I can't contact you because you haven't provided any email address)
